File: packages/jake_flutter/ios/Classes/JakeFlutterPlugin.swift
import Flutter
import JakeSDK
import UIKit

@MainActor
public final class JakeFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, JakeDelegate {
  private var eventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = JakeFlutterPlugin()
    let methods = FlutterMethodChannel(
      name: "ai.tryjake.sdk/methods",
      binaryMessenger: registrar.messenger()
    )
    let events = FlutterEventChannel(
      name: "ai.tryjake.sdk/events",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: methods)
    events.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      switch call.method {
      case "configure":
        try configure(call.arguments)
        result(nil)
      case "authenticate":
        guard
          let arguments = call.arguments as? [String: Any],
          let userId = arguments["userId"] as? String,
          let token = arguments["token"] as? String
        else {
          throw BridgeError("invalid_authentication", "userId and token are required.")
        }
        Task { @MainActor in
          do {
            try await Jake.authenticate(userId: userId, token: token)
            result(nil)
          } catch {
            result(flutterError(error))
          }
        }
      case "present":
        try Jake.presentMessenger()
        result(nil)
      case "dismiss":
        Jake.dismiss()
        result(nil)
      case "logout":
        Jake.logout()
        result(nil)
      case "track":
        guard
          let arguments = call.arguments as? [String: Any],
          let event = arguments["event"] as? String
        else {
          throw BridgeError("invalid_event", "event is required.")
        }
        Jake.track(
          event,
          properties: try values(arguments["properties"] as? [String: Any] ?? [:])
        )
        result(nil)
      case "setUserAttributes":
        Jake.setUserAttributes(try values(call.arguments as? [String: Any] ?? [:]))
        result(nil)
      case "setPushToken":
        guard
          let token = call.arguments as? String,
          let data = Data(hex: token)
        else {
          throw BridgeError(
            "invalid_push_token",
            "The APNs token must be a hexadecimal string."
          )
        }
        Jake.setPushToken(data)
        result(nil)
      case "getUnreadCount":
        result(Jake.unreadCount)
      default:
        result(FlutterMethodNotImplemented)
      }
    } catch {
      result(flutterError(error))
    }
  }

  public func onListen(
    withArguments _: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    eventSink = events
    Jake.delegate = self
    return nil
  }

  public func onCancel(withArguments _: Any?) -> FlutterError? {
    eventSink = nil
    Jake.delegate = nil
    return nil
  }

  public func jakeDidUpdateUnreadCount(_ count: Int) {
    eventSink?(["type": "unreadCountChanged", "count": count])
  }

  public func jakeAuthenticationDidExpire() {
    eventSink?(["type": "authenticationExpired"])
  }

  public func jakeDidFail(with error: JakeError) {
    eventSink?([
      "type": "error",
      "code": error.bridgeCode,
      "message": error.localizedDescription,
    ])
  }

  private func configure(_ rawArguments: Any?) throws {
    guard
      let arguments = rawArguments as? [String: Any],
      let workspaceId = arguments["workspaceId"] as? String,
      !workspaceId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
      let publicKey = arguments["publicKey"] as? String,
      !publicKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    else {
      throw BridgeError(
        "invalid_configuration",
        "workspaceId and publicKey are required."
      )
    }
    let messengerURL: URL
    if let value = arguments["messengerUrl"] as? String {
      guard
        let parsed = URL(string: value),
        let scheme = parsed.scheme?.lowercased(),
        ["http", "https"].contains(scheme)
      else {
        throw BridgeError(
          "invalid_configuration",
          "messengerUrl must use HTTP or HTTPS."
        )
      }
      messengerURL = parsed
    } else {
      messengerURL = JakeConfiguration.hostedMessengerURL
    }
    try Jake.configureMessenger(
      workspaceId: workspaceId,
      publicKey: publicKey,
      messengerURL: messengerURL
    )
  }

  private func values(_ dictionary: [String: Any]) throws -> [String: JakeValue] {
    try dictionary.mapValues { try JakeValue(bridgeValue: $0) }
  }

  private func flutterError(_ error: Error) -> FlutterError {
    if let bridge = error as? BridgeError {
      return FlutterError(code: bridge.code, message: bridge.message, details: nil)
    }
    let jake = error as? JakeError
    return FlutterError(
      code: jake?.bridgeCode ?? "jake_error",
      message: error.localizedDescription,
      details: nil
    )
  }
}

private struct BridgeError: LocalizedError {
  let code: String
  let message: String

  init(_ code: String, _ message: String) {
    self.code = code
    self.message = message
  }

  var errorDescription: String? { message }
}

private extension JakeError {
  var bridgeCode: String {
    switch self {
    case .notConfigured: "not_configured"
    case .invalidConfiguration: "invalid_configuration"
    case .invalidAuthentication: "invalid_authentication"
    case .authenticationRequired: "authentication_required"
    case .tokenStorageFailed: "token_storage_failed"
    case .presentationUnavailable: "presentation_unavailable"
    case .messengerLoadFailed: "messenger_load_failed"
    }
  }
}

private extension JakeValue {
  init(bridgeValue value: Any) throws {
    switch value {
    case is NSNull: self = .null
    case let value as Bool: self = .boolean(value)
    case let value as NSNumber:
      let double = value.doubleValue
      self = double.rounded() == double ? .integer(value.intValue) : .double(double)
    case let value as String: self = .string(value)
    default:
      throw BridgeError(
        "invalid_value",
        "Properties must be strings, numbers, booleans, or null."
      )
    }
  }
}

private extension Data {
  init?(hex: String) {
    let normalized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
    guard normalized.count.isMultiple(of: 2) else { return nil }
    var data = Data()
    var index = normalized.startIndex
    while index < normalized.endIndex {
      let next = normalized.index(index, offsetBy: 2)
      guard let byte = UInt8(normalized[index..<next], radix: 16) else { return nil }
      data.append(byte)
      index = next
    }
    self = data
  }
}

File: packages/jake_flutter/android/src/main/kotlin/ai/tryjake/flutter/JakeFlutterPlugin.kt
package ai.tryjake.flutter

import android.app.Activity
import android.os.Handler
import android.os.Looper
import ai.tryjake.sdk.Jake
import ai.tryjake.sdk.JakeEvent
import ai.tryjake.sdk.JakeEventListener
import ai.tryjake.sdk.JakeException
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class JakeFlutterPlugin :
  FlutterPlugin,
  MethodChannel.MethodCallHandler,
  EventChannel.StreamHandler,
  ActivityAware,
  JakeEventListener {
  private lateinit var methods: MethodChannel
  private lateinit var events: EventChannel
  private lateinit var applicationContext: android.content.Context
  private var activity: Activity? = null
  private var eventSink: EventChannel.EventSink? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    applicationContext = binding.applicationContext
    methods = MethodChannel(binding.binaryMessenger, "ai.tryjake.sdk/methods")
    events = EventChannel(binding.binaryMessenger, "ai.tryjake.sdk/events")
    methods.setMethodCallHandler(this)
    events.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methods.setMethodCallHandler(null)
    events.setStreamHandler(null)
    Jake.removeListener(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    try {
      when (call.method) {
        "configure" -> {
          val workspaceId = call.argument<String>("workspaceId").orEmpty()
          val publicKey = call.argument<String>("publicKey").orEmpty()
          Jake.configure(
            applicationContext,
            workspaceId,
            publicKey,
            call.argument<String>("messengerUrl") ?: Jake.DEFAULT_MESSENGER_URL,
          )
          result.success(null)
        }
        "authenticate" -> {
          Jake.authenticate(
            call.argument<String>("userId").orEmpty(),
            call.argument<String>("token").orEmpty(),
          )
          result.success(null)
        }
        "present" -> {
          Jake.present(
            activity ?: throw JakeException(
              "presentation_unavailable",
              "No foreground Activity is available.",
            ),
          )
          result.success(null)
        }
        "dismiss" -> {
          Jake.dismiss()
          result.success(null)
        }
        "logout" -> {
          Jake.logout()
          result.success(null)
        }
        "track" -> {
          Jake.track(
            call.argument<String>("event").orEmpty(),
            call.argument<Map<String, Any?>>("properties").orEmpty(),
          )
          result.success(null)
        }
        "setUserAttributes" -> {
          @Suppress("UNCHECKED_CAST")
          Jake.setUserAttributes(call.arguments as? Map<String, Any?> ?: emptyMap())
          result.success(null)
        }
        "setPushToken" -> {
          Jake.setPushToken(call.arguments as? String ?: "")
          result.success(null)
        }
        "getUnreadCount" -> result.success(Jake.unreadCount)
        else -> result.notImplemented()
      }
    } catch (error: JakeException) {
      result.error(error.code, error.message, null)
    } catch (error: Throwable) {
      result.error("jake_error", error.message, null)
    }
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
    eventSink = events
    Jake.addListener(this)
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
    Jake.removeListener(this)
  }

  override fun onEvent(event: JakeEvent) {
    mainHandler.post {
      when (event) {
        is JakeEvent.UnreadCountChanged -> eventSink?.success(
          mapOf("type" to "unreadCountChanged", "count" to event.count),
        )
        is JakeEvent.AuthenticationExpired -> eventSink?.success(
          mapOf("type" to "authenticationExpired"),
        )
        is JakeEvent.Error -> eventSink?.success(
          mapOf("type" to "error", "code" to event.code, "message" to event.message),
        )
      }
    }
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    activity = null
  }
}
